common4j/src/main/com/microsoft/identity/common/java/authorities/NativeAuthCIAMAuthority.kt

```diff
@@ -0,0 +1,112 @@
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+package com.microsoft.identity.common.java.authorities
+
+import com.microsoft.identity.common.java.exception.ClientException
+import com.microsoft.identity.common.java.logging.LogSession
+import com.microsoft.identity.common.java.logging.Logger
+import com.microsoft.identity.common.java.providers.nativeauth.NativeAuthOAuth2Configuration
+import com.microsoft.identity.common.java.providers.nativeauth.NativeAuthOAuth2Strategy
+import com.microsoft.identity.common.java.providers.nativeauth.NativeAuthOAuth2StrategyFactory
+import com.microsoft.identity.common.java.providers.oauth2.OAuth2StrategyParameters
+
+// TODO risk: this Authority class is not composable through Authority.getAuthorityFromAuthorityUrl()
+//  which is the method that's used throughout the project to create Authorities.
+class NativeAuthCIAMAuthority (
+    private val authorityUrl: String,
+    val clientId: String
+) : CIAMAuthority(authorityUrl) {
+    companion object {
+        private val TAG = NativeAuthCIAMAuthority::class.java.simpleName
+        private val NATIVE_AUTH_USE_OPENID_CONFIGURATION = false
+
+        @Throws(Exception::class)
+        fun getAuthorityFromAuthorityUrl(authorityUrl: String, clientId: String):
+                NativeAuthCIAMAuthority {
+            // Piggy back on the existing authority creation to improve reliability.
+            val authority = Authority.getAuthorityFromAuthorityUrl(authorityUrl)
+
+            // Check what authority was returned (must be CIAM or Native Auth CIAM)
+            return if (authority !is CIAMAuthority) {
+                throw ClientException(ClientException.NATIVE_AUTH_INVALID_CIAM_AUTHORITY)
+            } else if (authority is CIAMAuthority && authority !is NativeAuthCIAMAuthority) {
+                // Authority returned was a base CIAM authority
+                NativeAuthCIAMAuthority(
+                    authorityUrl = authority.authorityUri.toString(),
+                    clientId = clientId
+                )
+            } else {
+                // Authority is already a NativeAuthCIAMAuthority
+                authority as NativeAuthCIAMAuthority
+            }
+        }
+    }
+
+
+    // TODO audience, slice, flight parameters, multiple clouds supported,
+    // isAuthorityHostValidationEnabled (AzureActiveDirectoryOAuth2Configuration). Consider extending
+    // AzureActiveDirectoryAuthority
```

Comment:
I have no idea what this todo and these couple lines of comments are meant to indicate. Please elaborate here

Reply:
@SammyO

Reply:
@shahzaibj @fadidurah we left this TODO because we wanted to discuss this with you. We need some help understanding this logic in Authority:
 - what is "multiple clouds" used for - are we safe to say it's `false` for us, as it's `false` for CIAMAuthority? 
 - what is slice (and `AzureActiveDirectorySlice`)? How is it used? 
 - what are flight parameters used for? Are they relevant to us? 
 - what is host validation used for, and what should it be for us? 

Basically these are some Azure authority specific things that we're unclear on and need some guidance on :)

Reply:
- Multiple clouds - if your app supports users living in different AAD clouds such as Worldwide, USGov etc.
- Slice and DC - This is for targeting your request to a specific scale unit in aad (if you want to test something deployed there)
- Flight Parameters - Don't have much experience with this myself but I suspect we might have added these to enable something for flighting. cc/ @mohitc1 
- Host validation - Not sure. Can you point me to where you are seeing this?

Reply:
@shahzaibj  `AzureActiveDirectoryOAuth2Configuration.isAuthorityHostValidationEnabled()`. 

@SaurabhMSFT let's remove this TODO comment. As per Shahzaib's explanation we don't need any of these things right now. We should create a ticket to remove our DC handling and use the existing MSAL's one (using `AzureActiveDirectorySlice`). I was not aware of this component up until now, so we created our own custom implementation.